evm/src/main/java/org/hyperledger/besu/evm/code/CodeV1.java

```diff
@@ -38,6 +573,216 @@ public class CodeV1 implements Code {
+        if (pos + 1 > size) {
+          return "Truncated jump table";
+        }
+        final int jumpTableSize = RelativeJumpVectorOperation.getVectorSize(code, pos);
+        if (jumpTableSize == 0) {
+          return "Empty jump table";
+        }
+        pcPostInstruction += 1 + 2 * jumpTableSize;
+        if (pcPostInstruction > size) {
+          return "Truncated jump table";
+        }
+        for (int offsetPos = pos + 1; offsetPos < pcPostInstruction; offsetPos += 2) {
+          final int offset = Words.readBigEndianI16(offsetPos, rawCode);
+          final int rjumpdest = pcPostInstruction + offset;
+          if (rjumpdest < 0 || rjumpdest >= size) {
+            return "Relative jump destination out of bounds";
+          }
+          rjumpdests.set(rjumpdest);
+        }
+      } else if (operationNum == CallFOperation.OPCODE || operationNum == JumpFOperation.OPCODE) {
+        if (pos + 2 > size) {
+          return "Truncated CALLF/JUMPF";
+        }
+        int section = (rawCode[pos] & 0xff) << 8 | (rawCode[pos + 1] & 0xff);
+        if (section >= sectionCount) {
+          return "CALLF/JUMPF to non-existent section - " + Integer.toHexString(section);
+        }
+        pcPostInstruction += 2;
+      }
+      immediates.set(pos, pcPostInstruction);
+      pos = pcPostInstruction;
+    }
+    if ((attribute & TERMINAL) != TERMINAL) {
+      return "No terminating instruction";
+    }
+    if (rjumpdests.intersects(immediates)) {
+      return "Relative jump destinations targets invalid immediate data";
+    }
+    return null;
+  }
+
+  static String validateStack(final EOFLayout eofLayout) {
+    return IntStream.range(0, eofLayout.getCodeSections().length)
+        .mapToObj(i -> validateStack(i, eofLayout.getCodeSections()))
+        .filter(Objects::nonNull)
+        .findAny()
+        .orElse(null);
+  }
+
+  /**
+   * Validates the stack heights per <a href="https://eips.ethereum.org/EIPS/eip-5450">EIP-5450</a>.
+   *
+   * <p>This presumes code validation has already been performed, so there are no RJUMPS into
+   * immediates as well as no immediates falling off of the end of code sections.
+   *
+   * @param codeSectionToValidate The index of code to validate in the code sections
+   * @param codeSections The code sections to use for CALLF and JUMPF reference validation.
+   * @return null if valid, otherwise an error string providing the validation error.
+   */
+  public static String validateStack(
+      final int codeSectionToValidate, final CodeSection[] codeSections) {
+    byte[] code = codeSections[codeSectionToValidate].code.toArrayUnsafe();
+    int codeLength = code.length;
+    int[] stackHeights = new int[codeLength];
+    Arrays.fill(stackHeights, -1);
+
+    int thisWork = 0;
+    int maxWork = 1;
+    int[][] workList = new int[codeLength][2];
+
+    int initialStackHeight = codeSections[codeSectionToValidate].getInputs();
+    int maxStackHeight = initialStackHeight;
```

Comment:
is `stackHighWater` a better name, since this isn't really a limit, but rather a running max of the stack depth we update over time?

Reply:
Reference code in the EIP uses `max_stack_height`, so I think the intent is clear enough.